matrix_registration/api.py

```diff
@@ -141,6 +141,18 @@ class RegistrationForm(Form):
     )
 
 
+def get_request_ips(request):
+    """
+    Get the chain of client and proxy IP addresses from the request as
+    a nonempty list, where the closest IP in the chain is last. Each
+    IP vouches only for the IP before it. This works best if all proxies
+    conform the to the X-Forwarded-For header spec, including whatever
+    reverse proxy (such as nginx) is directly in front of the app, if any.
+    (X-Real-IP and similar are not supported at this time.)
+    """
+    return request.headers.getlist('X-Forwarded-For') + [request.remote_addr]
```

Comment:
```suggestion
    return request.headers.getlist('X-Forwarded-For')
```
see https://github.com/ZerataX/matrix-registration/issues/82#issuecomment-997547253

Reply:
It's important to keep remote_addr here because we don't know if nginx (or another _trusted_ proxy) is the closest hop.

Imagine the situation `1.2.3.4 -> 5.6.7.8 -> nginx -> app`, where both 5.6.7.8 and nginx are passing XFF headers and nginx shows up as 1.0.0.1 to the app. In this case, nginx passes `X-Forwarded-For: 1.2.3.4, 5.6.7.8` to the app and the app sees `remote_addr = 1.0.0.1`. The result is `['1.2.3.4', '5.6.7.8', '1.0.0.1']`.

Now take away nginx, producing a different situation the app has to handle: `1.2.3.4 -> 5.6.7.8 -> app`. Now the app directly receives `X-Forwarded-For: 1.2.3.4` and `remote_addr = 5.6.7.8`. The result is `['1.2.3.4', '5.6.7.8']`. If `remote_addr` weren't attached here, we'd miss the most important IP address, 5.6.7.8 (since it could be lying about forwarding on behalf of 1.2.3.4.)

Reply:
You're right I was forgetting about the possiblity for non trusted proxies!

Reply:
But since most of the users will use a reverse proxy you could check if the last entry in the array is the same as request.get_remote_address to keep the db clean
```suggestion
    ips = request.headers.getlist('X-Forwarded-For') or ["127.0.0.1"]
    ips += [request.get_remote_address] if not ips[-1] == request.get_remote_address
    return ips
```
(pls check the syntax im writing python not very frequent)

Reply:
That wouldn't help the case where there are no proxies in the picture, e.g. `1.2.3.4 -> app`. Here, we have an empty (missing) X-Forwarded-For and `remote_addr = 1.2.3.4`, so the returned result is `['1.2.3.4']`.

One improvement could be to remove the last item from the list if it's in a private IP range, but I don't think it's worth it.

Reply:
(No need for the `or`, by the way -- `getlist` will return `[]` if the header is missing.)

Reply:
...sorry, for some reason I missed the last two lines of your suggested diff. But I think with your suggestion you'd end up with entries like `127.0.0.1, 1.2.3.4` when it should just be `1.2.3.4`.

(Also, remember that the remote address as seen by the app without a reverse proxy is the same as the remote address that the reverse proxy would see if there was one.)

Reply:
> ...sorry, for some reason I missed the last two lines of your suggested diff. But I think with your suggestion you'd end up with entries like `127.0.0.1, 1.2.3.4` when it should just be `1.2.3.4`.

didn't you say that the or part doesn't even trigger?

> (Also, remember that the remote address as seen by the app without a reverse proxy is the same as the remote address that the reverse proxy would see if there was one.)

then the if part shouldn't trigger, this is only to not end up having the proxies address in there twice if its already in xff from using the updated config you provided

Reply:
There should be no situation where the proxy's address is included twice. Each address is only observed and reported by the next node in the chain, not by itself.

Reply:
Example: 
user `1.2.3.4` ["1.2.3.4"]
user's proxy `2.3.4.5` (does not use $proxy_add_x_forwarded_for but $remote_addr)  ["2.3.4.5"] 
reverse-proxy `10.0.0.1` (does use $proxy_add_x_forwarded_for) ["2.3.4.5", "10.0.0.1"]
python code (xff+get_remote_address) ["2.3.4.5", "10.0.0.1", "10.0.0.1"]

Reply:
checked the syntax again the suggest should have been
```suggestion
    return set(request.headers.getlist('X-Forwarded-For') + [request.get_remote_address])
```
which is way shorter and works just as well because set only has unique values and join works as well

Reply:
In your example of 1.2.3.4 -> 2.3.4.5 -> reverse-proxy -> app, here's what I see happening:

1. User at 1.2.3.4 sends request, no XFF
2. Proxy at 2.3.4.5 receives request. It sees a remote address of 1.2.3.4 and sticks that into a new XFF header. The request it sends along contains `X-Forwarded-For: 1.2.3.4`
3. Reverse proxy receives request from IP 2.3.4.5, so it appends that to the XFF. The request it sends contains `X-Forwarded-For: 1.2.3.4, 2.3.4.5`.
4. The app now sees the request from 10.0.0.1 with header `X-Forwarded-For: 1.2.3.4, 2.3.4.5` and puts that together as `1.2.3.4, 2.3.4.5, 10.0.0.1`.

You also really don't want to use a set here—the order is critical. When I refer to this as a chain, it's a chain of trust. The app can only vouch for having received the request from 10.0.0.1; 10.0.0.1 can only vouch for having received the request from 2.3.4.5; 2.3.4.5 can only vouch for having received the request from 1.2.3.4. Any one of these nodes could be lying! 2.3.4.5 could lie and send an XFF of 8.8.8.8 and we'd have no way of knowing. The way to read the resulting chain is by starting at the right and walking left until you reach the first IP you don't trust. But only the operator can make that trust decision, so we have to leave it to them.

Reply:
Sry was tired and didn't read correctly:
I just missunderstood $proxy_add_x_forwarded_for which adds the incoming ip to the xff chain, not its own...
You still need to switch to the new flask version tho (#83)
```suggestion
    return request.headers.getlist('X-Forwarded-For') + [request.get_remote_address]
```

Reply:
I think that deprecation only applies to flask-limiter, which dropped `get_ipaddr` in 2.0.1. As far as I can tell, in Flask itself, `request.remote_addr` is still the correct property.

Reply:
kk, i was confused because I couldn't find it in the flask doc because [it's accually part of werkzeug](https://werkzeug.palletsprojects.com/en/2.0.x/wrappers/#werkzeug.wrappers.Request.remote_addr) and thus failed to seperate it from pr-83